### ISSUE TITLE
BlockRenderContext requires RenderType

### ISSUE THREAD

**Issue by AViewFromTheTop** [FiniteReality/embeddium#284]
### Bug Description

Hello!
I am the lead developer of Wilder Wild.
Upon looking at me.jellysquid.mods.sodium.client.render.chunk.compile.pipeline.BlockRenderContext, I found that the class differs from that of Sodium's.
This parity is crucial as I am importing both Embeddium and Sodium to support custom Liquid and Snowlogging rendering, and Snowlogging specifically requires a mixin that calls the update method in this class. However, since they share the same exact file paths, Embeddium overrides Sodium's source, ensuring one of the two mods will be guaranteed to crash with Wilder Wild.

### Reproduction Steps

Compare Sodium's and Embeddium's sources.

### Log File

[latest.log](https://github.com/embeddedt/embeddium/files/15206442/latest.log)

**Comment by embeddedt** [comment]
We can remove this parameter for 1.20.x on Fabric as it's not used and we do intend to keep binary compatibility with Sodium 0.5.8 as closely as possible. However, it will probably get added back in a future version of MC to keep the Forge & Fabric branches more in sync.

That said, as mentioned on the [developer page](https://github.com/embeddedt/embeddium/wiki/For-Developers#how-to-add-integration) we prefer that people don't mixin to add custom functionality, as it means we can never change any of these functions without breaking some mod. Can you explain in more detail what you are trying to accomplish?

**Comment by AViewFromTheTop** [comment]
Rendering Snow Layers inside of other blocks.
I can’t add custom models because a lot of plants have offsets in-game.

**Comment by AViewFromTheTop** [comment]
![Screenshot_20240504_113721](https://github.com/embeddedt/embeddium/assets/87103914/08cd67c5-8319-46ab-9002-0bbd0d36c43d)
This should give a better view of what's going on.

**Comment by AViewFromTheTop** [comment]
If there's a way to maintain Sodium's structure while also supporting Embeddium's structure, that would be preferable.
I just have no way of importing two sources with the same paths, let alone compiling for both of them.
Do you think it would be possible to create a second method in the class that uses Sodium's parameters, and passes a dummy RenderLayer if possible? I'm not too sure how that would play out.

**Comment by AViewFromTheTop** [comment]
Quick update, Treetrain1 found an API you guys have!
He’s going to test it out to see if that fixes the issue or not.

**Comment by embeddedt** [comment]
> If there's a way to maintain Sodium's structure while also supporting Embeddium's structure, that would be preferable.
I just have no way of importing two sources with the same paths, let alone compiling for both of them.

We're going to fix the method signature to match Sodium's regardless of my comment about mixins, as it's a valid parity issue, and we intend to maintain binary compatibility with mods targeting Sodium 0.5.8 where possible. (This of course won't apply to Sodium 0.6+.)

> Rendering Snow Layers inside of other blocks.
I can’t add custom models because a lot of plants have offsets in-game.

I assume this is what Treetrain1 found but our [`ChunkMeshEvent`](https://github.com/embeddedt/embeddium/blob/20.1/forge/src/main/java/org/embeddedt/embeddium/api/ChunkMeshEvent.java) should be usable for this purpose.

**Comment by AViewFromTheTop** [comment]
Alright, thank you!
I wasn’t able to get the API working anyway.

**Comment by embeddedt** [comment]
Is this helpful? I realized the API is not too intuitive without having already studied the vanilla meshing code in detail :sweat_smile:, so I'll probably add a similar example somewhere in the codebase for other modders.

```java
public static void onChunkMesh(ChunkMeshEvent event) {
        LevelChunk chunk = event.getWorld().getChunk(event.getSectionOrigin().x(), event.getSectionOrigin().z());
        LevelChunkSection section = chunk.getSection(chunk.getSectionIndexFromSectionY(event.getSectionOrigin().y()));
        // Never schedule a mesh appender for empty sections (unless you need to render geometry in them).
        // Embeddium normally greatly speeds up initial world loads by not scheduling a rebuild task for an empty
        // section (since there is nothing to build), but the task is still created if you register a mesh appender
        // because mods may wish to render custom geometry in otherwise empty sections.
        // Since we only want to render a snowlogged effect on actual blocks, we don't need to bother with empty
        // sections.
        if(section != null && !section.hasOnlyAir()) {
            BlockState snowState = Blocks.SNOW.defaultBlockState().setValue(SnowLayerBlock.LAYERS, 7);
            event.addMeshAppender(ctx -> {
                // This part of the code runs on the meshing thread, not the main thread
                BlockPos.MutableBlockPos cursor = new BlockPos.MutableBlockPos();
                var vertexConsumer = ctx.vertexConsumerProvider().apply(RenderType.solid());
                PoseStack stack = new PoseStack();
                RandomSource random = RandomSource.create();
                // Even though we have a snapshot of the world, we still use the same coordinates
                int baseX = ctx.sectionOrigin().minBlockX();
                int baseY = ctx.sectionOrigin().minBlockY();
                int baseZ = ctx.sectionOrigin().minBlockZ();
                for(int y = 0; y < 16; y++) {
                    for(int z = 0; z < 16; z++) {
                        for(int x = 0; x < 16; x++) {
                            cursor.set(baseX + x, baseY + y, baseZ + z);
                            // Access blocks through the context's getter, not the section itself, as the latter
                            // is not safe to use on a meshing thread
                            BlockState selfState = ctx.blockRenderView().getBlockState(cursor);
                            if(!selfState.isAir()) {
                                // Check if the block underneath is netherite block
                                cursor.move(0, -1, 0);
                                if(ctx.blockRenderView().getBlockState(cursor).is(Blocks.NETHERITE_BLOCK)) {
                                    // Snowlog the original block
                                    cursor.move(0, 1, 0);
                                    stack.pushPose();
                                    stack.translate(x, y, z);

                                    // FIXME horrible hack to hide z-fighting of the sides of the snow
                                    float scaleFactor = 0.99f;
                                    float translateFactor = (1 - scaleFactor) / 2;
                                    stack.translate(translateFactor, translateFactor, translateFactor);
                                    stack.scale(scaleFactor, scaleFactor, scaleFactor);
                                    
                                    // Render the snow blockstate at the given position
                                    Minecraft.getInstance().getBlockRenderer().renderBatched(snowState, cursor, ctx.blockRenderView(), stack, vertexConsumer, true, random);
                                    
                                    // Reset the PoseStack for the next use
                                    stack.popPose();
                                }
                            }
                        }
                    }
                }
            });
        }
}
```

**Comment by AViewFromTheTop** [comment]
> Is this helpful? I realized the API is not too intuitive without having already studied the vanilla meshing code in detail 😅, so I'll probably add a similar example somewhere in the codebase for other modders.
> 
> ```java
> public static void onChunkMesh(ChunkMeshEvent event) {
>         LevelChunk chunk = event.getWorld().getChunk(event.getSectionOrigin().x(), event.getSectionOrigin().z());
>         LevelChunkSection section = chunk.getSection(chunk.getSectionIndexFromSectionY(event.getSectionOrigin().y()));
>         // Never schedule a mesh appender for empty sections (unless you need to render geometry in them).
>         // Embeddium normally greatly speeds up initial world loads by not scheduling a rebuild task for an empty
>         // section (since there is nothing to build), but the task is still created if you register a mesh appender
>         // because mods may wish to render custom geometry in otherwise empty sections.
>         // Since we only want to render a snowlogged effect on actual blocks, we don't need to bother with empty
>         // sections.
>         if(section != null && !section.hasOnlyAir()) {
>             BlockState snowState = Blocks.SNOW.defaultBlockState().setValue(SnowLayerBlock.LAYERS, 7);
>             event.addMeshAppender(ctx -> {
>                 // This part of the code runs on the meshing thread, not the main thread
>                 BlockPos.MutableBlockPos cursor = new BlockPos.MutableBlockPos();
>                 var vertexConsumer = ctx.vertexConsumerProvider().apply(RenderType.solid());
>                 PoseStack stack = new PoseStack();
>                 RandomSource random = RandomSource.create();
>                 // Even though we have a snapshot of the world, we still use the same coordinates
>                 int baseX = ctx.sectionOrigin().minBlockX();
>                 int baseY = ctx.sectionOrigin().minBlockY();
>                 int baseZ = ctx.sectionOrigin().minBlockZ();
>                 for(int y = 0; y < 16; y++) {
>                     for(int z = 0; z < 16; z++) {
>                         for(int x = 0; x < 16; x++) {
>                             cursor.set(baseX + x, baseY + y, baseZ + z);
>                             // Access blocks through the context's getter, not the section itself, as the latter
>                             // is not safe to use on a meshing thread
>                             BlockState selfState = ctx.blockRenderView().getBlockState(cursor);
>                             if(!selfState.isAir()) {
>                                 // Check if the block underneath is netherite block
>                                 cursor.move(0, -1, 0);
>                                 if(ctx.blockRenderView().getBlockState(cursor).is(Blocks.NETHERITE_BLOCK)) {
>                                     // Snowlog the original block
>                                     cursor.move(0, 1, 0);
>                                     stack.pushPose();
>                                     stack.translate(x, y, z);
> 
>                                     // FIXME horrible hack to hide z-fighting of the sides of the snow
>                                     float scaleFactor = 0.99f;
>                                     float translateFactor = (1 - scaleFactor) / 2;
>                                     stack.translate(translateFactor, translateFactor, translateFactor);
>                                     stack.scale(scaleFactor, scaleFactor, scaleFactor);
>                                     
>                                     // Render the snow blockstate at the given position
>                                     Minecraft.getInstance().getBlockRenderer().renderBatched(snowState, cursor, ctx.blockRenderView(), stack, vertexConsumer, true, random);
>                                     
>                                     // Reset the PoseStack for the next use
>                                     stack.popPose();
>                                 }
>                             }
>                         }
>                     }
>                 }
>             });
>         }
> }
> ```

I’m not entirely sure to be honest
My main issue with it was that it still required me to use that method with RenderLayer, which was impossible with both Embeddium and Sodium sources imported.
Perhaps if there was a method in an API that ran the update method, that could fix it? Because renderLayer can be gotten from the block context if I remember correctly, so a method that does that is viable.
It would also be nice for the model shaper to be passed as a parameter, but it’s not needed.

**Comment by AViewFromTheTop** [comment]
In any case once I get to my PC, how should I use this?
Like would would I get the chunk mesh event?
It would probably fix it honestly.

**Comment by embeddedt** [comment]
> Perhaps if there was a method in an API that ran the update method, that could fix it? Because renderLayer can be gotten from the block context if I remember correctly, so a method that does that is viable.

The idea of the API is that you do your extra rendering using vanilla functions, not Embeddium functions (as the latter are intended for internal use). We really don't want mods needing to do much special-casing for Embeddium, beyond some safer equivalent of mixing into the chunk renderer.

For instance, in the example above I call the vanilla block model renderer to add the snow layer. Aside from the use of the API hook to get the vertex consumer for the chunk section, there's nothing Embeddium-specific about that rendering logic - it uses existing vanilla infrastructure. This example doesn't use the `BlockRenderContext` at all.

> Like would would I get the chunk mesh event?

You can register an event handler with `ChunkMeshEvent.BUS.addListener`. It takes a `Consumer<ChunkMeshEvent>` as a parameter, so you can pass a lambda, method reference, or whatever.

**Comment by embeddedt** [comment]
For the record, I am just about to release 0.3.18 with the method signature change (writing the changelog as we speak), so your existing Sodium mixin will probably work anyway. The API is intended as our long-term solution, as we're not able to retain compatibility with Sodium after they release 0.6, and we're trying to wean mods off of using mixins where we can. In short - if you don't want to look into this right now, it may not be required in the short term.

**Comment by AViewFromTheTop** [comment]
Ah, that makes sense. I’ll try to add that in! Thank you!
Also, is there anything in particular going on with Sodium 0.6+ for you all to make that change? Or is it just a separate decision?

**Comment by embeddedt** [comment]
> is there anything in particular going on with Sodium 0.6+ for you all to make that change? Or is it just a separate decision?

I'll try to keep this concise to minimize potential drama from the answer and also to be respectful of your time, as the full answer would be rather long. :laughing: 

We (Nolij and myself, the current two maintainers of Embeddium) have had some disagreements with CaffeineMC in the past over various concerns we had with Sodium's project management (such as breaking changes which were not strictly necessary and were very likely to break mods).

The main problem at this point, though, is that as of Sodium 0.6, the license was changed to a new, more restrictive one that is not open-source (it explicitly prevents competing forks from being published using the code). Since we obviously compete with Sodium, we can't pull anything from newer versions.

We personally do not want to use versions of Sodium under that new license, and we still think our approach to mod compatibility is ultimately more stable for modders, so we're continuing to maintain Embeddium as an independent project, under the original open-source license. For now, we're trying our best to retain compatibility with anything that depended on Sodium 0.5.8 (hence the method signature revert). In the future, we'll inevitably diverge since the projects will go in different directions.

**Comment by AViewFromTheTop** [comment]
Just curious how anyone’s expected to maintain compat with Sodium if they’re not open source 😭

**Comment by embeddedt** [comment]
> how anyone’s expected to maintain compat with Sodium if they’re not open source 😭

Everyone else can read the code (and even incorporate portions of it) as long as they don't write a mod deemed to be "competing" under the license. The license's terms are targeted directly at competing projects like ours, not any arbitrary software project.

For reference the terms of the license can be found [here](https://polyformproject.org/licenses/shield/1.0.0/); it doesn't really give a strict definition of what constitutes competing, so I will not comment on that myself either.

**Comment by AViewFromTheTop** [comment]
Ah

**Comment by AViewFromTheTop** [comment]
Oh, really quick, by the way!
Would you consider adding an API for Cloud positions as well?
Wilder Wild makes their position change with wind, and can be seen here: https://github.com/FrozenBlock/WilderWild/blob/master/src/main/java/net/frozenblock/wilderwild/mixin/client/embeddium/CloudRendererMixin.java
Would be nice for it to have a small API as well- the less mixins, the better in my opinion

**Comment by embeddedt** [comment]
I may be missing something but at first glance it looks like you could use `ModifyArg` to modify the camera X/Z passed to `LevelRenderer#renderClouds` from `LevelRenderer#render` (or modify the matrices themselves, if the camera position isn't enough), and it would work with Sodium out of the box without requiring a dedicated mixin for that code as well.

**Comment by AViewFromTheTop** [comment]
https://github.com/FrozenBlock/WilderWild/commit/b7444fdac9db0a792ba6c8316f565994c9a073b7
This is awesome thank you!!!!
Pink Petals no longer Z-Fight with snowlogging too :)

**Comment by AViewFromTheTop** [comment]
> I may be missing something but at first glance it looks like you could use `ModifyArg` to modify the camera X/Z passed to `LevelRenderer#renderClouds` from `LevelRenderer#render` (or modify the matrices themselves, if the camera position isn't enough), and it would work with Sodium out of the box without requiring a dedicated mixin for that code as well.

Hadn't thought of that.
I'll try to look into it, thank you!

**Comment by AViewFromTheTop** [comment]
Dang, the wind idea worked too! Thank you so much!!!
The only thing left honestly would be.... this monstrosity.
https://github.com/FrozenBlock/WilderWild/blob/master/src/main/java/net/frozenblock/wilderwild/mixin/client/embeddium/FluidRendererMixin.java
It's meant to render a block as a Liquid, using a static 16x16 texture.
I wouldn't be surprised if this was one that can't really be fixed/worked around with Embeddium or Vanilla methods, but I just wanted to bring it up in the event there was something that could aid it even a small bit.

**Comment by embeddedt** [comment]
That mixin actually doesn't seem to function with Embeddium in its current state, not sure why.

Anyway, from trying it in vanilla, it seems you just want to override the texture/color that's used for fluid rendering; nothing else gets changed. Since the mesoglea block can only ever contains water fluid states, you should be able to just register an override for the water fluid handler using Fabric's fluid API and conditionally return different textures. (This should work with Sodium/Embeddium out of the box.)

Here is an example, which makes fluids rendered on top of netherite blocks use the diamond block texture. It should be easy to adjust this to detect the block at the given position being mesoglea. 

The reason for doing it "weirdly" with the resource listener is to make it properly compose with other mods that also try to wrap water rendering. To be completely transparent, I'm not 100% sure if this is the proper way to wrap water rendering, but the Fabric Discord should be able to confirm if you have questions. The documentation does suggest that wrapping vanilla fluids is an intended use case.

```java
ResourceManagerHelper.get(PackType.CLIENT_RESOURCES).registerReloadListener(new SimpleSynchronousResourceReloadListener() {
            @Override
            public ResourceLocation getFabricId() {
                return new ResourceLocation("embeddium", "fluid_example");
            }

            FluidRenderHandler customWaterHandler;

            @Override
            public void onResourceManagerReload(ResourceManager resourceManager) {
                if(customWaterHandler == null) {
                    FluidRenderHandler originalHandler = FluidRenderHandlerRegistry.INSTANCE.get(Fluids.WATER);
                    // Assert that the original handler exists now, otherwise the crash will happen later inside our
                    // handler
                    Objects.requireNonNull(originalHandler);
                    customWaterHandler = new FluidRenderHandler() {
                        final TextureAtlasSprite[] diamondWaterSprites = new TextureAtlasSprite[3];

                        private boolean isSingleTexture(@Nullable BlockAndTintGetter view, @Nullable BlockPos pos) {
                            return view != null && pos != null && view.getBlockState(pos.below()).is(Blocks.NETHERITE_BLOCK);
                        }

                        @Override
                        public TextureAtlasSprite[] getFluidSprites(@Nullable BlockAndTintGetter view, @Nullable BlockPos pos, FluidState state) {
                            if(isSingleTexture(view, pos)) {
                                return diamondWaterSprites;
                            } else {
                                return originalHandler.getFluidSprites(view, pos, state);
                            }
                        }

                        @Override
                        public void reloadTextures(TextureAtlas textureAtlas) {
                            originalHandler.reloadTextures(textureAtlas);
                            var diamondBlock = textureAtlas.getSprite(new ResourceLocation("block/diamond_block"));
                            diamondWaterSprites[0] = diamondBlock; // still
                            diamondWaterSprites[1] = diamondBlock; // flowing
                            diamondWaterSprites[2] = diamondBlock; // overlay
                        }

                        // Delegate all other methods to the original

                        @Override
                        public void renderFluid(BlockPos pos, BlockAndTintGetter world, VertexConsumer vertexConsumer, BlockState blockState, FluidState fluidState) {
                            originalHandler.renderFluid(pos, world, vertexConsumer, blockState, fluidState);
                        }

                        @Override
                        public int getFluidColor(@Nullable BlockAndTintGetter view, @Nullable BlockPos pos, FluidState state) {
                            return isSingleTexture(view, pos) ? 0xFFFFFFFF : originalHandler.getFluidColor(view, pos, state);
                        }
                    };
                    FluidRenderHandlerRegistry.INSTANCE.register(Fluids.WATER, Fluids.FLOWING_WATER, customWaterHandler);
                }
                
                // Load textures for our handler
                customWaterHandler.reloadTextures(Minecraft.getInstance().getModelManager().getAtlas(TextureAtlas.LOCATION_BLOCKS));
            }
        });
```

**Comment by AViewFromTheTop** [comment]
Thank you very much!!
Is there a way to get the texture from the waterlogged block? Or does it have to be registered per-block?

**Comment by AViewFromTheTop** [comment]
![2024-05-07_12 10 14](https://github.com/embeddedt/embeddium/assets/87103914/35e351ff-399d-4c6d-9350-14408ea9ee93)
![Screenshot_20240507_121051](https://github.com/embeddedt/embeddium/assets/87103914/2e3fc486-de4a-419b-9f62-c0b07d4a3c23)
It doesn't seem to work. Being the dunce I am sometimes, it could just be me screwing something up on accident.

**Comment by embeddedt** [comment]
Let me check 1.20.6, I tested that example on 1.20.1 and Fabric has changed how fluid handler registration is done internally since then.

**Comment by embeddedt** [comment]
This revised example works on 1.20.6. `renderFluid` should not be overriden anymore (it causes the given sprites to be ignored), and injecting the handler from a resource reload listener is no longer required.

```java
        FluidRenderHandler originalHandler = FluidRenderHandlerRegistry.INSTANCE.get(Fluids.WATER);
        // Assert that the original handler exists now, otherwise the crash will happen later inside our
        // handler
        Objects.requireNonNull(originalHandler);
        var customWaterHandler = new FluidRenderHandler() {
            final TextureAtlasSprite[] diamondWaterSprites = new TextureAtlasSprite[3];

            private boolean isSingleTexture(@Nullable BlockAndTintGetter view, @Nullable BlockPos pos) {
                return view != null && pos != null && view.getBlockState(pos.below()).is(Blocks.NETHERITE_BLOCK);
            }

            @Override
            public TextureAtlasSprite[] getFluidSprites(@Nullable BlockAndTintGetter view, @Nullable BlockPos pos, FluidState state) {
                if(isSingleTexture(view, pos)) {
                    return diamondWaterSprites;
                } else {
                    return originalHandler.getFluidSprites(view, pos, state);
                }
            }

            @Override
            public void reloadTextures(TextureAtlas textureAtlas) {
                originalHandler.reloadTextures(textureAtlas);
                var diamondBlock = textureAtlas.getSprite(new ResourceLocation("block/diamond_block"));
                diamondWaterSprites[0] = diamondBlock; // still
                diamondWaterSprites[1] = diamondBlock; // flowing
                diamondWaterSprites[2] = diamondBlock; // overlay
            }

            // Delegate all other methods to the original

            @Override
            public int getFluidColor(@Nullable BlockAndTintGetter view, @Nullable BlockPos pos, FluidState state) {
                return isSingleTexture(view, pos) ? 0xFFFFFFFF : originalHandler.getFluidColor(view, pos, state);
            }
        };
        FluidRenderHandlerRegistry.INSTANCE.register(Fluids.WATER, Fluids.FLOWING_WATER, customWaterHandler);
```

**Comment by AViewFromTheTop** [comment]
![2024-05-07_12 38 50](https://github.com/embeddedt/embeddium/assets/87103914/4f8eee95-6fd9-4dde-bffa-2bc573500012)
Huge step in the right direction, thank you!
Only thing now is that the sides are zoomed in, is there anything that can be done about that?
It's alright if not, I can just stick to my mixins I've been using.

**Comment by embeddedt** [comment]
> It's alright if not, I can just stick to my mixins I've been using.

I would *really* prefer not to have such an invasive mixin in the fluid renderer, because the mixin will inevitably break if I ever have to refactor it.

> Only thing now is that the sides are zoomed in, is there anything that can be done about that?

To solve that, I think you would either have to make another texture that's scaled the same way as the vanilla water fluid textures (so that it doesn't look zoomed in), *or* you could just conditionally take over fluid rendering and render it yourself in `renderFluid` with the desired textures (which might be preferable, if you already have the code for it for vanilla).

**Comment by AViewFromTheTop** [comment]
> I would _really_ prefer not to have such an invasive mixin in the fluid renderer, because the mixin will inevitably break if I ever have to refactor it.

I've updated it for Sodium's from 1.19 through now, but it's understandable!

> To solve that, I think you would either have to make another texture that's scaled the same way as the vanilla water fluid textures (so that it doesn't look zoomed in), _or_ you could just conditionally take over fluid rendering and render it yourself in `renderFluid` with the desired textures (which might be preferable, if you already have the code for it for vanilla).

I think that renderFluid one is my best bet. I made a method for this exact situation in Vanilla for the library WW uses!

**Comment by AViewFromTheTop** [comment]
![2024-05-07_12 57 11](https://github.com/embeddedt/embeddium/assets/87103914/9cd003e6-7483-46c8-a93a-5aff0c6930d3)
Awesome to know that code still works with Embeddium/Sodium.
Thank you so much!!!!!

**Comment by embeddedt** [comment]
Double check that Sodium actually respects the `renderFluid` hook in 0.5.8. I get a feeling that's one of the many promised bugfixes which have been sitting on the development branch unable to be released for months. So you might need to still use your original mixin if you detect Sodium and not Embeddium.

**Comment by AViewFromTheTop** [comment]
Will do!

**Comment by AViewFromTheTop** [comment]
![2024-05-07_13 01 06](https://github.com/embeddedt/embeddium/assets/87103914/508fa864-0551-44d3-a310-40c2a7c65405)
Yeah... mixin it is for Sodium.

**Comment by embeddedt** [comment]
Is https://github.com/FrozenBlock/WilderWild/blob/master/src/main/java/net/frozenblock/wilderwild/mixin/client/embeddium/FluidRendererMixin.java still required?

**Comment by AViewFromTheTop** [comment]
> Is https://github.com/FrozenBlock/WilderWild/blob/master/src/main/java/net/frozenblock/wilderwild/mixin/client/embeddium/FluidRendererMixin.java still required?

For a *very* specific use case, possibly so. It’s to make the bottom side of Water render as the solid-colored overlay.

**Comment by embeddedt** [comment]
What triggers that? When I look up at the underside of ocean water with just WW installed I still see the regular still texture.

**Comment by AViewFromTheTop** [comment]
> What triggers that? When I look up at the underside of ocean water with just WW installed I still see the regular still texture.

I meant the bottom of the block, not the underside of the surface.
Without this change, placing blocks like glass underwater looks very annoying in my opinion, especially with a full glass box. (The sides would use the solid overlay, but the ceilings show the ripple texture- with this, it’s all solid)

**Comment by embeddedt** [comment]
I see. That doesn't seem to be working with just WW though.

![image](https://github.com/embeddedt/embeddium/assets/42941056/dec32b50-11bb-4082-b6d5-a1909d4dae82)

Anyway, this type of change is probably another good candidate to be done conditionally in `renderFluid`.

**Comment by AViewFromTheTop** [comment]
Would you be able to try that scenario without WW
The water face on the top part of the glass should have the ripple pattern as well without it

**Comment by AViewFromTheTop** [comment]
On that note though, I’m not really sure how to easily do this using renderFluid without pretty much overwriting all water rendering
I find the mixin simpler because it just swaps out the bottom texture instead of overwriting things

**Comment by embeddedt** [comment]
Oh, now I see what you mean; the difference is extremely subtle to my eyes though (as the ripple is nowhere near as obvious as it is on the bottom of the glass).

> I find the mixin simpler because it just swaps out the bottom texture instead of overwriting things

The mixin is easy for you to write, but it's very much not easy for me when players complain about it being broken down the line. :wink: That problem doesn't exist for vanilla mixins because the code is obviously not changing until another Minecraft update, but cross-mod mixins tend to be rather unstable.

This would be the kind of niche visual change where if you don't want to have the responsibility of replacing the water renderer using `renderFluid`, I'd be tempted to suggest "just have this not work with Embeddium", as I'm frankly not sure most users would notice the difference, and I'm having trouble thinking of a generic enough API that would be useful for more than just this specific example. `renderFluid` is Fabric's standard approach for deep customization of how a fluid renders.

I'll keep thinking and let you know if I do come up with something though.

**Comment by AViewFromTheTop** [comment]
I feared it was too niche for that 😭

**Comment by embeddedt** [comment]
Going to close this as the original issue was fixed in 0.3.18, but feel free to reach out if you need anything else!

**Comment by AViewFromTheTop** [comment]
> Going to close this as the original issue was fixed in 0.3.18, but feel free to reach out if you need anything else!

Thank you very much!!!
